cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -541,6 +541,115 @@ type ResetConfiguration struct {
 	Timeouts *Timeouts
 }
 
+// UpgradeApplyConfiguration contains a list of configurable options which are specific to the "kubeadm upgrade apply" command.
+type UpgradeApplyConfiguration struct {
```

Comment:
i realized we are missing one important option here
`kubernetesVersion`

currently that's an required argument for the `apply` command.
do we want to have it in upgradeApplyConfiguration?

my initial reaction is yes.
but it might be a bit tricky to make the CLI arg override the field in config.

Reply:
It would be good to have.

> but it might be a bit tricky to make the CLI arg override the field in config.

As we will print the version in the upgrade progress, this will be clear to user, IMO.

Reply:
ok, let's add it.

> If the field and arg are different, we can print a warning log.

i think we should try first to override the field with the arg.
if not so easy to do, let's error saying `errors.Errorf("the provided upgrade version in the configuration file is different: %s`, val)

Reply:
The `diff` command also seems miss the option, I wonder whether the `required` parameter is suitable for a configuration file?

Reply:
ok, I agree with you.

Reply:
> The diff command also seems miss the option, I wonder whether the required parameter is suitable for a configuration file?

we should have it in all upgrade configs that need version. in the ClusterConfiguration we do have required fields like kubernetesVersion.